examples/transactions_example/processor.go

```diff
@@ -0,0 +1,354 @@
+			istate.lightState[road] = "green"
+		} else {
+			istate.lightState[road] = "red"
+		}
+
+		isectMsg.Lights[road] = lightStateMsg{
+			Name:        istate.name,
+			Road:        road,
+			State:       istate.lightState[road],
+			CarsWaiting: istate.carsWaiting[road]}
+	}
+
+	value, err := json.Marshal(isectMsg)
+	if err != nil {
+		fatal(err)
+	}
+
+	err = producer.Produce(
+		&kafka.Message{
+			TopicPartition: kafka.TopicPartition{
+				Topic:     &outputTopic,
+				Partition: kafka.PartitionAny,
+			},
+			Key:   []byte(istate.name),
+			Value: value,
+		}, nil)
+
+	if err != nil {
+		fatal(fmt.Sprintf("Failed to produce message: %v", err))
+	}
+
+	return changed
+}
+
+func trafficLightProcessor(wg *sync.WaitGroup, termChan chan bool) {
+	defer wg.Done()
+
+	doTerm := false
+	ticker := time.NewTicker(500 * time.Millisecond)
+
+	// If punctuate is true the current intersection state is
+	// written to the output topic every ticker interval regardless if
+	// there was a intersection light state change.
+	// Use this to get a more responsive visualization.
+	punctuate := true
+
+	intersectionStates = make(map[string]*intersectionState)
+
+	// The per-partition producers are set up in groupRebalance
+	producers = make(map[int32]*kafka.Producer)
+
+	consumerConfig := &kafka.ConfigMap{
+		"client.id":         "processor",
+		"bootstrap.servers": brokers,
+		"group.id":          processorGroupID,
+		"auto.offset.reset": "earliest",
+		// Consumer used for input to a transactional processor
+		// must have auto-commits disabled since offsets
+		// are committed with the transaction using
+		// SendOffsetsToTransaction.
+		"enable.auto.commit":     false,
+		"go.logs.channel.enable": true,
+		"go.logs.channel":        logsChan,
+	}
+
+	var err error
+	processorConsumer, err = kafka.NewConsumer(consumerConfig)
+	if err != nil {
+		fatal(err)
+	}
+
+	err = processorConsumer.Subscribe(inputTopic, groupRebalance)
+	if err != nil {
+		fatal(err)
+	}
+
+	addLog(fmt.Sprintf("Processor: waiting for messages on topic %s",
+		inputTopic))
+	for !doTerm {
+		select {
+
+		case <-ticker.C:
+			// Run intersection state machine(s) every second.
+			partitionsToCommit := make(map[int32]bool)
+			for _, istate := range intersectionStates {
+				if intersectionStateMachine(istate) || punctuate {
+					// The state machine wants its transaction committed.
+					// The transaction is shared among all intersectionStates
+					// that use the same input partition since the input
+					// offset that is committed along with the transaction
+					// applies to all intersectionStates mapped to
+					// that partition.
+					partitionsToCommit[istate.partition] = true
+				}
+			}
+
+			// Commit transactions
+			for partition := range partitionsToCommit {
+				commitTransactionForInputPartition(partition)
+			}
+
+		case <-termChan:
+			doTerm = true
+
+		default:
+			// Poll consumer for new messages or rebalance events.
+			ev := processorConsumer.Poll(100)
```

Comment:
seems fine, but would it be a bit better to have poll have a timeout of `0` and have the ticker interval drive everything?

Reply:
We want to consume messages as quickly as possible, blocking on the ticker would only consume when the ticker fires, which I think is atypical.